cpp/include/cudf/reduction.hpp

```diff
@@ -69,6 +73,44 @@ std::unique_ptr<scalar> reduce(
   data_type output_dtype,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief Compute reduction to each list in the lists column.
+ *
+ * This function does not detect overflows in reductions.
+ * Using a higher precision `data_type` may prevent overflow.
+ * The null values are skipped for the operation.
+ * If the list is empty, the row corresponding to the result of the
+ * list is null.
+ *
+ * @throw cudf::logic_error if reduction is called for non-arithmetic output
+ * type and operator other than `min` and `max`.
+ * @throw cudf::logic_error if input column data type is not convertible to
+ * output data type.
+ * @throw cudf::logic_error if `min` or `max` reduction is called and the
+ * output type does not match the input column data type.
+ * @throw cudf::logic_error if `any` or `all` reduction is called and the
+ * output type is not bool8.
+ *
+ * If the input column has arithmetic type, output_dtype can be any arithmetic
+ * type. If the input column has non-arithmetic type, e.g. timestamp or string,
+ * the same output type must be specified.
+ *
+ * @param col Input lists column.
+ * @param agg Aggregation operator applied by the reduction.
+ * @param output_dtype  The computation and output precision.
+ * @param null_handling If `INCLUDE`, the reduction is valid if all elements in
+ * a segment are valid, otherwise null. If `EXCLUDE`, the reduction is valid if
+ * any element in the segment is valid, otherwise null.
+ * @param mr Device memory resource used to allocate the returned scalar's device memory.
+ * @returns Output column with results of segmented reduction.
+ */
+std::unique_ptr<column> segmented_reduce(
+  lists_column_view const& col,
```

Comment:
Remind me why this is a `lists_column_view` instead of separate offsets/data? This isn't really a "segmented" reduction anymore. It's a "per-list reduction".

Reply:
Good question! I briefly alluded to this during our libcudf meeting today but didn’t fully explain due to time. If you recall the conversations we had about indexing and iterators on #9588, there are two styles of indexing used in libcudf. One is what I call “N+1”, and requires the indices to be monotonic and contiguous. The endpoint of one segment is the start of the next segment. This is how lists and strings columns are indexed. The other style is what I call “2N”, where the start and endpoints are independent, and the segments can be arbitrarily overlapping. That is used for slicing, and is the format used by cub (which we call to perform the reduction).

The 2N indexing is most powerful when the index iterators are template parameters, allowing for arbitrary device iterators as we enabled for null mask reductions in #9588. We don’t want a template function in the public API, but that would be our preferred internal/detail implementation for maximum flexibility and alignment with the cub code that we call.

After discussion with @isVoid, we were unable to think of a second use case for the public API accepting a data column and a column of N+1 indices. The only use case we could identify was the list reduction, so a stronger commitment to the N+1 indexing scheme in a public API seemed pointless. Moreover, a public API accepting N+1 indices would need to be bounds-checked to ensure safety while the public API we settled on for a lists column reduction does not need to be checked since we know the list indexing is contiguous and monotonic by construction.

Reply:
Should we rename this API to `lists_reduce`?

Reply:
> The 2N indexing is most powerful when the index iterators are template parameters

Agreed, luckily with transform iterators its pretty easy to translate `2N` <-> `N+1`. 

> we were unable to think of a second use case for the public API accepting a data column and a column of N+1 indices

Doing segmented reductions is quite normal in C++, so it only makes sense to me expose it in a canonical/generic way that can be used both for lists and just standard segmented reductions.

Reply:
Okay. Then these are the steps I'd suggest:
- API is `segmented_reduce(column_view data, column_view indices, ...)`, where `indices` are of the N+1 format
- Implement bounds-checking over indices to ensure they are monotonic/contiguous (we also need to document those requirements but we have to bounds-check anyway because we don't want to allow illegal memory accesses from public APIs)
- Consider whether to add a convenience method for `lists_reduce(lists_column_view, ...)`

Future PR:
- Refactor the detail/internal method to use 2N indexing with iterator template parameters

Reply:
I would do:
```
segmented_reduce(column_view data, device_span<size_type> offsets, ...)
```

Likewise, you can have two overloads of `segmented_reduce` for `2N` and `N+1` similar to `split/slice`. 

> Implement bounds-checking over indices to ensure they are monotonic/contiguous (we also need to document those requirements but we have to bounds-check anyway because we don't want to allow illegal memory accesses from public APIs)

https://github.com/rapidsai/cudf/issues/5505

Reply:
Changed to `device_span` at public API level for indices. Per last two comments above.

Reply:
I've chatted with @bdice offline and we both thought making a `device_span` from a `column_view` is not particularly easy today. We were thinking of a convenience function for user. At one point we were thinking of `to_device_span` member method; but on a second thought, but perhaps a `make_device_span` factory function is better?

Reply:
FYI Essentially the inverse operations is requested in #9656.

Reply:
Just add an implicit conversion operation to `column_view` to `device_span<T>` that verifies the type of the `column_view` matches `T`.

Reply:
In progress in #10302.

Reply:
#10302 is now merged. To wrap up the previous pieces of the discussion:
- We will not implement bounds checking, but rather document the expectations of the API per #5505.
- We will not add a convenience method for `lists_reduce(lists_column_view, ...)` unless requested later. For now, @revans2 is fine with the APIs available in this PR and #10302.
- Follow-up work will refactor the detail/internal method to use 2N indexing with iterator template parameters.
I think that resolves this conversation.